### PR TITLE
Add basic auth support for Prometheus pushgateway

### PR DESCRIPTION
**Pull request by AndrewDi**
We hide pushgateway behind nginx, and use HTTP basic auth, so need pushgateway to implement it.

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@AndrewDi Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=22548)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-boot&pullRequestId=22548) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@AndrewDi Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=22548)!

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java [hunk 1]
```diff
@@ -116,6 +118,21 @@ public Pushgateway getPushgateway() {
 		 */
 		private Map<String, String> groupingKey = new HashMap<>();
 
+		/**
+		 * Enable publishing via a Prometheus Pushgateway with Basic Auth.
+		 */
+		private Boolean authEnabled = false;
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java:124]: We don't have such `enabled` flag. What we usually do is to check that the username is set, something like `StringUtils.hasText` can help you with that.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java [hunk 2]
```diff
@@ -116,6 +118,21 @@ public Pushgateway getPushgateway() {
...
+
+		/**
+		 * Basic Auth Username.
+		 */
+		private String authUsername;
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java:129]: We usually use a plain `username` for this.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java [hunk 3]
```diff
@@ -116,6 +118,21 @@ public Pushgateway getPushgateway() {
 		 */
 		private Map<String, String> groupingKey = new HashMap<>();
 
+		/**
+		 * Enable publishing via a Prometheus Pushgateway with Basic Auth.
+		 */
+		private Boolean authEnabled = false;
+
+		/**
+		 * Basic Auth Username.
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java:127]: I am not sure if "basic auth" is really important here.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java [hunk 4]
```diff
@@ -86,6 +86,8 @@ public Pushgateway getPushgateway() {
 		return this.pushgateway;
 	}
 
+
+
```
Comment (snicoll) [inline spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java:90]: Please remove this.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java (hunk 2 above)
Comment (AndrewDi) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java:127]: @snicoll  all lower case?

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java (hunk 3 above)
Comment (AndrewDi) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java:125]: pushgateway already support basic auth, we just need to pass auth parameters to `setConnectionFactory` function.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java (hunk 2 above)
Comment (snicoll) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java:127]: Yes. As I mentioned, this is what we usually do so you should find a number of similar cases searching the codebase.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java (hunk 3 above)
Comment (snicoll) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java:125]: I am aware of that. The comment is on the key description.

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java (hunk 4 above)
Comment (AndrewDi) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java]: done

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java (hunk 2 above)
Comment (AndrewDi) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java:127]: @snicoll I'm aware of that, but I found other variables like `baseUrl`\`pushRate`\`groupingKey`, key word is uppper case?

Comment (snicoll) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java:127]: > you should find a number of similar cases searching the codebase.

All properties that represents a username in this codebase are defined as `username`. So, yes, as described in my first comment.

Comment (AndrewDi) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java:127]: @snicoll  so you advise `authusername` and `authPassword` ?

Comment (AndrewDi) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java:127]: a little wired

Comment (snicoll) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java:127]: No. As mentioned in my first comment, `username` (and `password`).

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java (hunk 1 above)
Comment (AndrewDi) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java:122]: you mean check this value in `PrometheusMetricsExportAutoConfigurationTests`?

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java (hunk 2 above)
Comment (AndrewDi) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java:127]: very well, done

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java (hunk 3 above)
Comment (AndrewDi) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java]: ok

File: spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java (hunk 1 above)
Comment (snicoll) [reply spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/metrics/export/prometheus/PrometheusProperties.java:122]: No. I mean that we don't enable authentication based on a flag but based on the fact that the `username` property is set.

**Comment by snicoll** [comment]
@AndrewDi how is it going? Did you get a chance to look at the review?

**Comment by snicoll** [comment]
I've decided to give this one anyway, see [this polish commit](https://github.com/spring-projects/spring-boot/commit/14eb041086f7696a8a57f6c9b91c9ee5086c580f).

**Pull request closed without merging**